plugins/flashrom/fu-flashrom-lspcon-i2c-spi-device.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (C) 2021 Daniel Campello <campello@chromium.org>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-flashrom-device.h"
+#include "fu-flashrom-lspcon-i2c-spi-device.h"
+
+#include <libflashrom.h>
+
+#define I2C_PATH_REGEX "/i2c-([0-9]+)/"
+#define HID_LENGTH 8
+#define DEVICE_GUID_FORMAT "FLASHROM-I2C\\VEN_%s&amp;DEV_%s"
+
+struct _FuFlashromLspconI2cSpiDevice {
+	FuFlashromDevice		 parent_instance;
+	gint				 bus_number;
+	guint8				 active_partition;
+};
+
+G_DEFINE_TYPE (FuFlashromLspconI2cSpiDevice, fu_flashrom_lspcon_i2c_spi_device,
+	       FU_TYPE_FLASHROM_DEVICE)
+
+struct romentry {
+	uint32_t start;
+	uint32_t end;
+	bool included;
+	const char *name;
+	const char *file;
+};
+
+struct flashrom_layout {
+	/* entries store the entries specified in a layout file and associated
+	 * run-time data */
+	struct romentry *entries;
+	/* the number of successfully parsed entries */
+	size_t num_entries;
+};
+
+static struct romentry entries[3] = {
+	{
+		.start = 0x00002,
+		.end = 0x00003,
+		.included = true,
+		.name = "FLAG"
+	},
+	{
+		.start = 0x10000,
+		.end = 0x1ffff,
+		.included = true,
+		.name = "PARTITION1"
+	},
+	{
+		.start = 0x20000,
+		.end = 0x2ffff,
+		.included = true,
+		.name = "PARTITION2"
+	},
+};
+
+static struct flashrom_layout layout = {entries, 3};
+
+#define VERSION_ADDR_OFFSET	0x5000
+
+
+static void
+fu_flashrom_lspcon_i2c_spi_device_init (FuFlashromLspconI2cSpiDevice *self)
+{
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_set_version_format (FU_DEVICE (self), FWUPD_VERSION_FORMAT_PAIR);
+}
+
+static gboolean
+fu_flashrom_lspcon_i2c_spi_device_prepare (FuDevice *device,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuFlashromDevice *parent = FU_FLASHROM_DEVICE (device);
+	FuFlashromLspconI2cSpiDevice *self = FU_FLASHROM_LSPCON_I2C_SPI_DEVICE (device);
+
+	/* flashrom_programmer_init() mutates the programmer_args string. */
+	fu_flashrom_device_set_programmer_args (parent,
+						g_strdup_printf	("bus=%d", self->bus_number));
+
+	/* FuFlashromDevice->prepare */
+	if (!FU_DEVICE_CLASS
+	    (fu_flashrom_lspcon_i2c_spi_device_parent_class)->prepare (device,
+								       FWUPD_INSTALL_FLAG_NONE,
+								       error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gint
+fu_flashrom_lspcon_i2c_spi_device_get_bus_number_from_path (const gchar *path,
+							    GError **error)
+{
+	gint bus_number = -1;
+	GRegex *regex = NULL;
+	GMatchInfo *info;
```

Comment:
both g_autoptr()

Reply:
If I use g_autoptr() do I need the g_match_info_free and g_regex_unref below?

Reply:
Nope, it does those automatically when the variable goes out of scope.